### DISCUSSION TITLE
Local authentication support

### DISCUSSION THREAD

**Discussion by kvaps** [ltsp/ltsp#303] [Q&A]
Hi,
We need an option for disable ltsp pam authentication.
Some users uses sssd and other methods for authentication.

Optionally we also need to keep local accounts as well. 

I would like to add two new variables:

- `LOCAL_AUTH=<0|1>` - for enable or disable local authentication (default: 0)
- `CLEANUP_USERS=<0|1>` - for enable or disable ltsp users cleanup (default: 1)

**Comment by kvaps** [comment]
One more question, how can I include `home/*` to squashed image without editing `/usr/share/ltsp/server/image/image.excludes` file?

**Comment by alkisg** [comment]
To override any function without introducing parameters and without providing 56-xxx.sh scripts, one can use lines like the following in ltsp.conf:

```shell
[server]
PRE_IMAGE_NO_REMOVE_USERS="eval 'remove_users() { echo Omitting remove_users; }'"

[clients]
PRE_INIT_NO_PAM_MAIN="eval 'pam_main() { echo Omitting pam_main; }'"
```

If that's considered ugly, we may introduce a new parameter for that, e.g.

```shell
OMIT_FUNCTIONS="remove_users pam_main"
```

To include home/* one way would be to clone image.excludes to /etc/ltsp while omitting that line. I'm not sure if mksquashfs has parameters to include files, if so, that would help. Ideas welcome.

**Comment by alkisg** [comment]
Maybe we can introduce a regex, e.g. NO_IMAGE_EXCLUDES="^(home|var)/", that would then call this:

```shell
grep -vE "$NO_IMAGE_EXCLUDES" image.excludes > $(mktemp -d)/image.excludes
```

...and then use the temp image.excludes as a parameter to mksquashfs.

**Comment by kvaps** [comment]
I like `OMIT_FUNCTIONS` way

>To include home/* one way would be to clone image.excludes to /etc/ltsp while omitting that line. 

This will not work, because both files are used, local and upstream one:
https://github.com/ltsp/ltsp/blob/master/ltsp/server/image/65-mksquashfs.sh#L30-L31

I think we have to add two more options for configure mksquashfs without creating own image.excludes file. Also consider to not use regex syntax, it might be much simpler:

```
NO_IMAGE_EXCLUDES="home/*"
ADD_IMAGE_EXCLUDES="
foo
bar/*
baz/sub
"
```

**Comment by alkisg** [comment]
> This will not work, because both files are used

No. There are two local files: image.excludes **overrides** the upstream one, while image-local.excludes is an additional one.

I like the ADD_IMAGE_EXCLUDES and OMIT_IMAGE_EXCLUDES approach. I think if those are implemented, we won't need image-local.excludes anymore, but it will still be nice to have the /etc/ltsp/image.excludes override in case someone wants something really customized.

Can you file a different issue or PR to discuss that part separately?
I'm not sure if we'll want to suggest multiline values by default, or if multiple variables would be better, e.g. ADD_IMAGE_EXCLUDES_FOO="foo", ADD_IMAGE_EXCLUDES_BAR="bar" etc.

And let's make this PR about implementing OMIT_FUNCTIONS (or close this one and open a new one for OMIT_FUNCTIONS).

**Comment by kvaps** [comment]
So after merging ltsp/ltsp#83 and ltsp/ltsp#84 local users can be enabled by specifying:
```ini
[server]
OMIT_FUNCTIONS="remove_users"
OMIT_IMAGE_EXCLUDES="home/*"
[clients]
OMIT_FUNCTIONS="pam_main"
```

**Comment by alkisg** [comment]
OK I moved this one to the community, to serve as a quick how-to, as the LTSP code changes that were required, were implemented as part of other issues.

Thanks!

**Comment by nwrox** [comment]
> So after merging [ltsp/ltsp#83](https://github.com/ltsp/ltsp/pull/83) and [ltsp/ltsp#84](https://github.com/ltsp/ltsp/pull/84) local users can be enabled by specifying:
> 
> ```ini
> [server]
> OMIT_FUNCTIONS="remove_users"
> OMIT_IMAGE_EXCLUDES="home/*"
> [clients]
> OMIT_FUNCTIONS="pam_main"
> ```

Hello, i've tried this with no luck, if i add the user to the server and run ```ltsp initrd```, when i reboot the client, the login screen starts blinking, i can see the user, but can't login (the screen "resets"). There is a way to keep the vm image as is (independent from the server)?

**Comment by alkisg** [comment]
Try to switch to vt2 and login from there. Note the error messages, if any.

Additionally, check the client /etc/passwd and /etc/shadow. Do they correctly list local users? See the [end of the ltsp.conf man page](https://ltsp.org/man/ltsp.conf/) for an example on how to set the root password to be able to login as root in the client.

That said, what's the use case? Where will /home come from in your scenario? Is it NFS, SSHFS, a local partition, a RAM disk?

**Comment by nwrox** [comment]
> Try to switch to vt2 and login from there. Note the error messages, if any.
>
> Additionally, check the client /etc/passwd and /etc/shadow. Do they correctly list local users? See the [end of the ltsp.conf man page]
(https://ltsp.org/man/ltsp.conf/) for an example on how to set the root password to be able to login as root in the client.
> 
> That said, what's the use case? Where will /home come from in your scenario? Is it NFS, SSHFS, a local partition, a RAM disk?

I must try this again, in my use case, i need to provide a clean linux env (debian + openbox) for thinclients to connect in our citrix env through citrix workspace app, but this isn't working so far.
